cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -92,7 +92,23 @@ type Etcd struct {
 	DataDir   string
 	ExtraArgs map[string]string
 	// Image specifies which container image to use for running etcd. If empty, automatically populated by kubeadm using the image repository and default etcd version
-	Image string
+	Image      string
+	SelfHosted *SelfHostedEtcd
+}
+
+// SelfHostedEtcd describes options required to configure self-hosted etcd
+type SelfHostedEtcd struct {
+	// CertificatesDir represents the directory where all etcd TLS assets are stored. By default this is
+	// a dir names "etcd" in the mains CertificatesDir value.
+	CertificatesDir string
+	// The service IP of the etcd cluster
```

Comment:
Wouldn't it be more convenient to require a service name instead of an IP and figure the IP out by using the k8s api?

Reply:
Or calculate this from ServiceSubnet

Reply:
Yeah, good idea. I'll use a new offset. I like the idea of using service name too, but I think it might be safer for the apiserver to rely on iptables rather than the dns server?